jenkins/PerfCI_Chaos/03_PerfCI_Choas_VMs_Deployment/Jenkinsfile

```diff
@@ -0,0 +1,197 @@
+
+                                # Get containers using the image
+                                containers=$(sudo podman ps -a --filter ancestor=${QUAY_BENCHMARK_RUNNER_REPOSITORY} -q)
+
+                                if [[ "$containers" != "" ]]; then
+                                    echo "Stopping and removing containers using the image..."
+                                    sudo podman stop $containers
+                                    sudo podman rm -f $containers
+                                fi
+
+                                # Now remove the image
+                                sudo podman rmi -f $(sudo podman images -q ${QUAY_BENCHMARK_RUNNER_REPOSITORY} 2> /dev/null)
+                            fi
+                        '''
+                }
+            }
+        }
+
+        stage('⚙️ SET SSH key') {
+            steps {
+                script {
+                    sh 'mkdir -p $WORKSPACE/.ssh/'
+                    sh 'sudo chmod 700 $WORKSPACE/.ssh/'
+                    withCredentials([file(credentialsId: 'perfci_provision_private_key_file', variable: 'PROVISION_PRIVATE_KEY_FILE')]) {
+                        sh "sudo cp ${PROVISION_PRIVATE_KEY_FILE} ${WORKSPACE}/.ssh/provision_private_key"
+                        sh "sudo chown jenkins:jenkins ${WORKSPACE}/.ssh/provision_private_key"
+                    }
+                    sh 'sudo chmod 600 $PRIVATE_KEY_PATH'
+                    sh '''
+                        sudo cat > "$CONFIG_PATH" <<END
+    Host provision
+        HostName ${PROVISION_IP}
+        User ${PROVISION_USER}
+        IdentityFile $PRIVATE_KEY_PATH
+        StrictHostKeyChecking no
+        ServerAliveInterval 30
+        ServerAliveCountMax 5
+END
+                    '''
+                    sh 'sudo chmod 600 $CONFIG_PATH'
+                }
+            }
+        }
+
+        stage('Prepare Google Credential Files') {
+            steps {
+                script {
+                    // Create a unique temp directory
+                    env.TEMP_DIR = sh(script: "umask 77; mktemp -d -p '${env.WORKSPACE_PATH}'", returnStdout: true).trim()
+                    if (!env.TEMP_DIR) {
+                        error "Failed to create temporary directory!"
+                    }
+
+                    // Set file paths
+                    env.GOOGLE_DRIVE_CREDENTIALS = "${env.TEMP_DIR}/${GOOGLE_CREDENTIALS_FILE}"
+                    env.GOOGLE_DRIVE_TOKEN = "${env.TEMP_DIR}/${GOOGLE_TOKEN_FILE}"
+
+                    // Prepare credential files
+                    sh """
+                        set -e
+                        umask 77
+                        cp '${GOOGLE_DRIVE_CREDENTIALS_FILE}' '${env.GOOGLE_DRIVE_CREDENTIALS}' || error "Failed to copy ${GOOGLE_DRIVE_CREDENTIALS_FILE}"
+                        cp '${GOOGLE_DRIVE_TOKEN_FILE}' '${env.GOOGLE_DRIVE_TOKEN}' || error "Failed to copy ${GOOGLE_DRIVE_TOKEN_FILE}"
```

Comment:
Do those variables get expanded by the shell or by Jenkins?  If the former they need to be double quoted; if the latter, this is OK.

Reply:
This variables get from Jenkins secret that handle by us

Reply:
What expands the variables, Jenkins or bash?

Reply:
bash

Reply:
GOOGLE_DRIVE_CREDENTIALS_FILE = credentials('perfci_google_drive_credentials_file')
GOOGLE_DRIVE_TOKEN_FILE = credentials('perfci_google_drive_token_file')
copy it from Jenkins secret to local Jenkins variable.